### PR TITLE
Youtube/Vimeo integration

### PR DESCRIPTION
**Pull request by jayoshih**
Dependent on https://github.com/learningequality/ricecooker/pull/45

### CODE REVIEW

File: ricecooker/classes/files.py [hunk 1]
```diff
@@ -100,23 +106,44 @@ def get_hash(filepath):
...
+        compress_video(config.get_storage_path(filename), tempf.name, overwrite=True, **ffmpeg_settings)
         filename = "{}.{}".format(get_hash(tempf.name), file_formats.MP4)
 
-        copy_file_to_storage(filename, tempf.name, delete_original=True)
+        copy_file_to_storage(filename, tempf.name)
+
+        FILECACHE.set(key, bytes(filename, "utf-8"))
+        return filename
+
+def download_from_web(web_url, download_settings):
```
Comment (jayoshih) [inline ricecooker/classes/files.py]: Code for downloading from youtube

File: ricecooker/classes/files.py [hunk 2]
```diff
@@ -24,6 +26,10 @@
 # Cache for filenames
 FILECACHE = FileCache(config.FILECACHE_DIRECTORY, forever=True)
 
+def generate_key(action, path_or_id, settings, default=" (default)"):
```
Comment (jayoshih) [inline ricecooker/classes/files.py]: Separated key generation for files with settings

Comment (aronasorman) [inline ricecooker/classes/files.py]: Can you provide examples of the parameters here?

File: ricecooker/classes/nodes.py [hunk 3]
```diff
@@ -5,31 +5,26 @@
 import sys
 from le_utils.constants import content_kinds,file_formats, format_presets, licenses, exercises
 from ..exceptions import InvalidNodeException, InvalidFormatException
-from .. import config
+from .. import config, __version__
 
-def guess_content_kind(files, questions=None):
+def guess_content_kind(path=None, web_video_data=None, questions=None):
```
Comment (aronasorman) [inline ricecooker/classes/nodes.py]: delete function

File: ricecooker/classes/nodes.py [hunk 4]
```diff
@@ -360,41 +360,39 @@ def process_files(self):
...
 
         downloaded = super(VideoNode, self).process_files()
 
-        try:
-            # Extract thumbnail if one hasn't been provided and derive_thumbnail is set
-            if self.derive_thumbnail and len(list(filter(lambda f: isinstance(f, ThumbnailFile), self.files))) == 0:
-                videos = list(filter(lambda f: isinstance(f, VideoFile), self.files))
-                assert len(videos) > 0 and videos[0].filename, "No videos downloaded for this node"
+        # Extract thumbnail if one hasn't been provided and derive_thumbnail is set
+        if self.derive_thumbnail and len(list(filter(lambda f: isinstance(f, ThumbnailFile), self.files))) == 0:
```
Comment (aronasorman) [inline ricecooker/classes/nodes.py]: change `len(list...()...) == 0` to:

```
if self.derive_thumbnail and any(f for f in self.files if isinstance(f, ThumbnailFile)):

```

File: ricecooker/classes/nodes.py [hunk 5]
```diff
@@ -360,41 +360,39 @@ def process_files(self):
...
         try:
             assert self.kind == content_kinds.VIDEO, "Assumption Failed: Node should be a video"
             assert self.license, "Assumption Failed: Video content must have a license"
             assert self.questions == [], "Assumption Failed: Video should not have questions"
             assert len(self.files) > 0, "Assumption Failed: Video must have at least one video file"
 
-            # Check if there are any .mp4 files
-            files_valid = False
+            # Check if there are any .mp4 files if there are video files (other video types don't have paths)
+            files_valid = len(set(filter(lambda f: isinstance(f, VideoFile), self.files))) == 0
```
Comment (aronasorman) [inline ricecooker/classes/nodes.py]: Change to `any`

**Pull request merged**